### ISSUE TITLE
TypeScript

### ISSUE THREAD

**Issue by LandonSchropp** [crswll/tailwindcss-theme-swapper#49]
Thank you for this plugin! 

I was wondering, have you considered converting the plugin to TypeScript or adding types to this repo to support Tailwind's TypeScript configuration?

![Screenshot 2024-01-01 at 10 50 30 AM](https://github.com/crswll/tailwindcss-theme-swapper/assets/361591/38eeeaa7-5b4a-4309-9058-3de06ae0fc3f)

**Comment by crswll** [comment]
No problem at all! I have considered adding types, even started a couple times but couldn't get it to be as intuitive as I want. I feel like [this is pretty close](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnA3nAwhAdgM2AHM4BfOfKCEOAchgENgAbAd2FwBMBjAZ19oAoQUlRwAIihRgACg1jAGzADwAVAHxwAvBkFw4AbRlwOcANYpEEfHFUBdAPwAuW0btwUADxgouvOBAARgBWKNzwDhJSsvIwiiqqbpouiTJ2ANyCpJmCHD5Q+AzcaADKrAxgqFDKAEIMvCiqABYoICia6Hr6CC1tvC4GXd3dncNj3bgMbS60gfUoQuPjDcxhMND9cACCUFAMiMq8MFAcROpDS-ptnIoAigCuKFCIznBHJ7hEF5cwvSgudQazVaKG+3VIABowfoAHRwnZ7A6jS4TKYoV7vU7QsYrNYbFwI-aHY6nc4o7rXO6PZ4YkmfbHDX4glySaRyBRKWrzYFtMlLUh8-R2LLCbh4I5wOYNbS6YZi5j42VLMAnEDyRAzKAoTi0KGXBpirjqlzI8mcFCFe7MGAzMUQZiLfkXUhZN4MOK8Qgofw4AjEAz0P60YWCQ0SpltMoVKoygAUmAj3oG7lILijlSeylEKGskvm6gAlNoOi6RH901V4xdE5tBpdTSjJtM6FKFnryW8UKtwoqA04qBAYMH2+TEy5W9jIUryU3-nQvFMwKtdQzurie1Ba7QYQvwMu7COUWPpx39PLFQ3T8MVaBjS3mI8V1eloFmAwmjNfsB-E8qJu4AwXA9O6Jj+EQEAQDqq7jC6p6wU6wzCqQBbCIIQA) but I don't really like having to do `Swapper<typeof baseTheme>`. Will take another crack at it soon.

**Comment by LandonSchropp** [comment]
I'm happy to jump in as well if you'd like. 🙂

Given the size of the repo, it might be easier to rewrite the source itself in TypeScript. Then the compiler generates the types for you and you don't have the extra maintenance burden of keeping the types in sync with the code.

I understand if you don't want to go down that route though.

**Comment by LandonSchropp** [comment]
I played with your config a bit and [here's what I came up with](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnA3nAwhAdgM2AHM4BfOfKCEOAchgENgAbAd2FwBMBjAZ19oAoQRxgoo+BtzQAVABYoQKHAWIZBcTXFwMlALji8YUDkQDcGrbxTMU3GNF4GjJ3EQDaAXQta4SzsAMAIoAruKIAPzOxqY+WjAK+th4hB70iSi0noKkwqLiktJwAEIM1vKKyinEADyl5RlwKAAeYly8yappCZVZAHxNrSjtcBVKKqnqvjpJAEQARmUosxa5IrhiElJoAMqsDGCoUBO19ShjaC1tnB0n3Rn9U-EZTnDulr6aZxd3dUsXfQANB9PgA6cE-aoebK+bJrbh4IxwRbWOAAXiemgRzEcBnQIN8YBMIAYUEQBloUBQnFowM+VjseE4pPJmPpWk4KEkIWYMApCIgzCE7LIINy5F4DBgwF4hBQtyh7nSvWyggRuCRPSUewOR3RcAAFJgtfKDF4yAYdYdxL8kKgIPhkUs+gBKdEDdBrE1Wo5GkEm17vEX4kXTXQoCkozJ00Oaay2ey4t60PRUCAwLIx2MmgxRglaUiAtkimYRugtXRgWy0-P0+N2BxQQO0UEV8DVzxZ0M54uxuDYpMhvsiomgFmR5hhGvDkXzZgMOQUhIyppQKhNuAMLgIORSuArogQCA02uh3LD8-sy+aOEu4SCIA). I wouldn't call myself a TypeScript expert though, and I'm not deeply familiar with the options in this repo, so I'd take that with a grain of salt.

**Comment by LandonSchropp** [comment]
In terms of preventing extra properties, I dug a bit and I don't think that's a thing TypeScript supports currently ([issue](https://github.com/microsoft/TypeScript/issues/12936)). It seems like it could be [possible](https://stackoverflow.com/a/57117594/262125) using some mangled type magic, but it doesn't seem worth it to me.

**Comment by crswll** [comment]
Yeah, it probably wouldn't hurt to write it in TypeScript. I mostly didn't because I didn't want to add a build step. It's likely worth it, though.

Your types are close but I think an important thing to make sure of is that the child themes can't declare theme values that aren't in the base theme which mine does but I don't think well. I'll start playing around with converting to TypeScript and maybe make some breaking changes I've wanted to.

**Comment by LandonSchropp** [comment]
Sounds good. 👍 

> the child themes can't declare theme values that aren't in the base theme

I don't think I quite understand this req. Maybe you could give a quick example. No worries if you just want to tackle it yourself as well—I don't mean to backseat drive. 🙂

**Comment by crswll** [comment]
Oh it's all good sorry if it sounds otherwise! Don't think you're backseat driving at all.

I mean like if the base theme has `colors: { a: '#f00', b: '#0f0' }` the following themes can only have `a` and/or `b` but not `c`.

**Comment by LandonSchropp** [comment]
Ah, I see what you were going for. I _think_ [this](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnA3nAwhAdgM2AHM4BfOfKCEOAchgENgAbAd2FwBMBjAZ19oAoQUlTY8hIgBUAFihBoAvOILEA2gCIYchRoC6w0WgAiKFGAAKDWMAbMAPFIB8cZekFw4ai3A5wA1iiIEPhwUnoA-ABcYd56cCgAHjAoXLxwEABGAFYo3PARcKbmVjZ2jnEuMVJxANyCpPWCHClQ+AzcaABCDLwosvIoOKpE9l0Jyamc6cOSAwouSSlpYTpDEsRjLu6enrgMCjEamb0oGvW7CGsxXQ3CLShtHWjz6yNjE8vTKnNr2x67faHOC8GBQDhEC67PrMPIwaC8GKg8G4IhqPRQzwKTi2ACKAFdHohoiCwRDMVdBjFipZrDBbA4uk47s1cK12p04ABlVgMMCoKCzTbjJZTGYbaR-DAAzzaQaIrwyy5wHp9V5C0ZMgA0SsuADoDeqJVt0UqDKRhAB6S1wACStBouAg8F4+KgaFYMgY8G0wHSfpB+P5ED6nAQEDg7pgbtwergAE0IPjaGHcGYw-C4JwUIQ05TIyho1BcAhkGhenBWChmMw9SIy6tBjy+QLXHAPqKVhrXk4ABTcCUxZv8x4arYASlcLgAbhBgJwmgPcKDKQphwLqmt1482-2JZPFNsLSIt7yR1BeztdnKFAq1LrdlflcqgSgYrQTn1aDrn8+YXCERiNRaCiKhnVoPQf1-S4bzfaVoOggdmEA+CELQsBwRAaxEHfd1OG-B80MuPol04bCYifIi0Ozdp8WYGB3wHCBmCEKjoItNjdg4ojuOfUgtVQtDX3fJIDjAWECKo-98hQ4C9VE8AJMgwjlVgiiVMQ5iUMozjPAw0ByLoTJmEJSTdMuYyGBkd9fXSR4qCgdIGC4K5vV8dIiAgCB8I039eIQ-zLn881xyAA) works. It uses generics to pass through the base theme config to the other types, ensuring that the other configs are always a deep partial of it.

The only downside of this approach I've found so far is that the error message isn't necessarily the easiest to understand.

**Comment by crswll** [comment]
Got started on this. Getting pretty tripped up using it with Tailwind's `plugin.withOption(...)` API so will need to figure that out. Definitely in the works though.

**Comment by crswll** [comment]
Had some time today and sit and work on this. I have to get the tooling set up around it but think the hard part is done. I still can't get the types passed properly to `plugin.withOptions<...>(plugin, extend)` and might just need to make the plugin a preset instead.

**Comment by FeliciaGartzLevin** [comment]
Hi! A junior dev here. I can't use the theme in a setup with TypeScript, Vite and React. I get the error: `[plugin:vite:css] [postcss] Cannot read properties of undefined (reading 'call')`, and in `tailwind.config.js` TS is asking me: 

> "Could not find a declaration file for module 'tailwindcss-theme-swapper'. '/node_modules/.pnpm/tailwindcss-theme-swapper@0.11.5_tailwindcss@3.4.3/node_modules/tailwindcss-theme-swapper/src/index.js'  implicitly has an 'any' type.
>   Try `npm i --save-dev @types/tailwindcss-theme-swapper` if it exists or add a new declaration (.d.ts) file containing `declare module 'tailwindcss-theme-swapper';`ts(7016)"

I read your conversation above. Does it mean that the plugin is not compatible at all with TypeScript, and is there a way to work around this? The types that you've tried and discussed above, in what file and where should it be added (if one would get it to work properly)?

Thanks in advance!
